File: printer.cpp
#include "printer.hpp"

#include <iostream>
#include <vector>

auto Printer::execute(std::vector<Expr*>& stmts) -> void {
    for (Expr* expr : stmts) {
        evaluate(expr);
        std::cout << "\n";
    }
}


// private methdos
auto Printer::evaluate(Expr* expr) -> Object* {
    return expr->accept(this);
}


// protected methods
auto Printer::visit_ternaryexpr(TernaryExpr* expr) -> Object* {
    std::cout << "( ";
    evaluate(expr->condition);
    std::cout << "? ";
    evaluate(expr->first);
    std::cout << ": ";
    evaluate(expr->second);
    std::cout << ") ";
    return nullptr;
}

auto Printer::visit_binaryexpr(BinaryExpr* expr) -> Object* {
    std::cout << "( " << expr->oper->lexeme << " ";
    evaluate(expr->left);
    evaluate(expr->right);
    std::cout << ") ";
    return nullptr;
}

auto Printer::visit_unaryexpr(UnaryExpr* expr) -> Object* {
    std::cout << "( " << expr->oper->lexeme << " ";
    evaluate(expr->expr);
    std::cout << ") ";
    return nullptr;
}

auto Printer::visit_exponentexpr(ExponentExpr* expr) -> Object* {
    std::cout << "( ** ";
    evaluate(expr->left);
    evaluate(expr->right);
    std::cout << ") ";
    return nullptr;
}

auto Printer::visit_literalexpr(LiteralExpr* expr) -> Object* {
    std::cout << expr->value->lexeme << " ";
    return nullptr;
}

auto Printer::visit_identifierexpr(IdentifierExpr* expr) -> Object* {
    std::cout << expr->name->lexeme << " ";
    return nullptr;
}

auto Printer::visit_groupingexpr(GroupingExpr* expr) -> Object* {
    std::cout << "( ";
    evaluate(expr->expr);
    std::cout << ") ";
    return nullptr;
}

auto Printer::visit_nilexpr([[maybe_unused]] NilExpr* expr) -> Object* {
    std::cout << "nil ";
    return nullptr;
}

File: lib/printer.ts
import IExpr, {
    BinaryExpr,
    ExponentExpr,
    GroupingExpr,
    IdentifierExpr,
    IExprVisitor,
    LiteralExpr,
    NilExpr,
    TernaryExpr,
    UnaryExpr,
} from "./expr"

export default class Printer implements IExprVisitor<void> {
    private constructor() {}

    static createInstance() {
        return new Printer()
    }

    execute(exprs: IExpr[]): void {
        exprs.forEach((expr) => {
            this.evaluate(expr)
            console.log()
        })
    }

    private evaluate(expr: IExpr) {
        expr.accept(this)
    }

    visitTernaryExpr(expr: TernaryExpr): void {
        process.stdout.write("( ")
        this.evaluate(expr.condition())
        process.stdout.write("? ")
        this.evaluate(expr.first())
        process.stdout.write(": ")
        this.evaluate(expr.second())
        process.stdout.write(") ")
    }
    visitBinaryExpr(expr: BinaryExpr): void {
        process.stdout.write("( ")
        this.evaluate(expr.left())
        process.stdout.write(expr.oper().lexeme() + " ")
        this.evaluate(expr.right())
        process.stdout.write(") ")
    }
    visitExponentExpr(expr: ExponentExpr): void {
        process.stdout.write("( ")
        this.evaluate(expr.left())
        process.stdout.write("** ")
        this.evaluate(expr.right())
        process.stdout.write(") ")
    }
    visitUnaryExpr(expr: UnaryExpr): void {
        process.stdout.write("( " + expr.oper().lexeme() + " ")
        this.evaluate(expr.expr())
    }
    visitGroupingExpr(expr: GroupingExpr): void {
        process.stdout.write("( ")
        this.evaluate(expr.expr())
        process.stdout.write(") ")
    }
    visitLiteralExpr(expr: LiteralExpr): void {
        process.stdout.write(expr.token().lexeme() + " ")
    }
    visitIdentifierExpr(expr: IdentifierExpr): void {
        process.stdout.write(expr.token().lexeme() + " ")
    }
    visitNilExpr(expr: NilExpr): void {
        process.stdout.write("nil ")
    }
}
